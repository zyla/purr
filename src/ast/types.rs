use super::{Located, QualifiedName, Symbol};
use crate::string::PSString;

pub type Type = Located<TypeKind>;

#[derive(Debug)]
pub enum TypeKind {
    Unknown(u64),
    Var(Symbol),
    /// Standalone operator
    Operator(Symbol),
    TypeLevelString(PSString),
    TypeLevelInt(num::BigInt),
    Wildcard(WildcardKind),
    TypeConstructor(QualifiedName),
    TypeApp(Box<Type>, Box<Type>),
    FunctionType(Box<Type>, Box<Type>),

    /// Kind parameter application. Doesn't seem to appear in source, because kind params are
    /// implicit?
    KindApp(Box<Type>, Box<Type>),

    ForAll {
        vars: Vec<(Symbol, Option<Box<Type>>)>,
        body: Box<Type>,

        // TODO: figure out whether this needs to be a part of vars
        skolem_scope: Option<SkolemScope>,
    },

    Constrained {
        constraint: Constraint,
        body: Box<Type>,
    },

    // TODO: I have no idea what these parameters mean
    Skolem(Symbol, Option<Box<Type>>, u64, SkolemScope),
    Row {
        fields: Vec<(Symbol, Type)>,
        rest: Option<Box<Type>>,
    },
    Kinded {
        r#type: Box<Type>,
        kind: Box<Type>,
    },
    Parens(Box<Type>),

    /// Infix operator sequence with unknown precedence
    Infix(Box<Type>, Vec<(Symbol, Type)>),
}

#[derive(Debug)]
pub enum WildcardKind {
    Hole(Symbol),
    Unnamed,
    Ignored,
}

#[derive(Debug)]
pub struct SkolemScope(pub u64);

pub type Constraint = Box<Type>;
/*
 * for now just Type
#[derive(Debug)]
pub struct Constraint {
    pub class: QualifiedName,
    pub kind_args: Vec<Type>,
    pub type_args: Vec<Type>,
    // TODO: PartialConstraintData - do we need it?
}
*/
