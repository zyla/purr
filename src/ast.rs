use crate::symbol::Symbol;

#[derive(Debug)]
pub struct SourceSpan {
    pub start: usize,
    pub end: usize,
}

#[derive(Debug)]
pub struct Located<T>(pub SourceSpan, pub T);

#[derive(Debug)]
pub struct Comment(pub String);

#[derive(Debug)]
pub struct Commented<T>(pub Vec<Comment>, pub T);

pub type Module = Located<Commented<ModuleInner>>;

#[derive(Debug)]
pub struct ModuleInner {
    pub name: Symbol,
    pub exports: Option<Vec<DeclarationRef>>,
    pub declarations: Vec<Declaration>,
}

#[derive(Debug)]
pub enum DeclarationRef {
    // TODO
}

pub type Declaration = Located<Commented<DeclarationKind>>;

#[derive(Debug)]
pub enum DeclarationKind {
    ValueDeclaration(ValueDeclaration),
}

#[derive(Debug)]
pub struct ValueDeclaration {
    pub ident: Symbol,

    // TODO: what is this?
    // pub name: NameKind

    // TODO
    // pub binders: Vec<Binder>
    pub expr: Vec<GuardedExpr>,
}

#[derive(Debug)]
pub struct GuardedExpr {
    pub guards: Vec<Guard>,
    pub expr: Expr,
}

#[derive(Debug)]
pub enum Guard {
    // TODO
}

pub type Expr = Located<ExprKind>;

#[derive(Debug)]
pub enum ExprKind {
    Literal(Literal<Expr>),
    // TODO
}

#[derive(Debug)]
pub enum Literal<T> {
    Integer(i32),
    Float(f64),
    String(String),
    Char(char),
    Boolean(bool),
    Array(Vec<T>),
    Object(Vec<(Symbol, T)>),
}
