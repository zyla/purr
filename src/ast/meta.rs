use salsa::DebugWithDb;

use crate::ModuleId;

use super::AbsoluteName;

#[derive(Eq, PartialEq, Debug, Hash, Clone, Copy, DebugWithDb)]
pub struct SourceSpan {
    pub decl: SpanDeclRef,
    pub start: usize,
    pub end: usize,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone, Copy, DebugWithDb)]
pub enum SpanDeclRef {
    Module(ModuleId),
    Decl(AbsoluteName),
    Unknown,
}

impl SourceSpan {
    pub fn new_in_module(start: usize, end: usize, module_id: ModuleId) -> Self {
        Self {
            start,
            end,
            decl: SpanDeclRef::Module(module_id),
        }
    }

    pub fn new_relative(start: usize, end: usize, name: AbsoluteName) -> Self {
        Self {
            start,
            end,
            decl: SpanDeclRef::Decl(name),
        }
    }

    pub fn to_relative(&mut self, name: AbsoluteName, reference_loc: usize) {
        self.decl = SpanDeclRef::Decl(name);
        self.start = self.start - reference_loc;
        self.end = self.end - reference_loc;
    }

    pub fn unknown() -> Self {
        // TODO: find a better representation?
        Self {
            start: 0,
            end: 0,
            decl: SpanDeclRef::Unknown,
        }
    }

    /// Alias for `unknown()`, but we can grep for it and gradually add sensible locations
    pub fn todo() -> Self {
        Self::unknown()
    }

    pub fn to_file_location(&self, _db: &dyn crate::Db) -> (String, usize, usize) {
        todo!()
    }
}

pub trait ToSourceSpan {
    fn to_source_span(&self, module_id: ModuleId) -> SourceSpan;
}

#[derive(Eq, PartialEq, Debug, Hash, Clone, DebugWithDb)]
pub struct Located<T>(pub SourceSpan, pub T);

impl<T> Located<T> {
    pub fn new(span: SourceSpan, x: T) -> Self {
        Self(span, x)
    }

    pub fn into_inner(self) -> T {
        self.1
    }

    pub fn span(&self) -> SourceSpan {
        self.0
    }

    pub fn to_relative_span(&mut self, name: AbsoluteName, reference_loc: usize) {
        self.0.to_relative(name, reference_loc)
    }
}

impl<T> std::ops::Deref for Located<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.1
    }
}

impl<T> std::ops::DerefMut for Located<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.1
    }
}

#[derive(Eq, PartialEq, Debug, Hash, Clone, DebugWithDb)]
pub struct Comment(pub String);

#[derive(Eq, PartialEq, Debug, Hash, Clone, DebugWithDb)]
pub struct Commented<T>(pub Vec<Comment>, pub T);

impl<T> Commented<T> {
    pub fn into_inner(self) -> T {
        self.1
    }
}

impl<T> std::ops::Deref for Commented<T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.1
    }
}
