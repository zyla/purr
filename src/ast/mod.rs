pub mod declarations;
pub mod expr;
pub mod meta;
pub mod types;

use crate::ModuleId;
pub use declarations::*;
pub use expr::*;
pub use meta::*;
pub use types::*;

use crate::symbol::Symbol;

#[salsa::interned]
pub struct AbsoluteName {
    pub module: ModuleId,
    pub name: Symbol,
}

#[salsa::interned]
pub struct QualifiedName {
    pub module: Option<ModuleId>,
    pub name: Symbol,
}

impl QualifiedName {
    pub fn is_actually_qualified(&self, db: &dyn crate::Db) -> bool {
        self.module(db).is_some()
    }

    pub fn new_unqualified(db: &dyn crate::Db, name: Symbol) -> Self {
        Self::new(db, None, name)
    }

    pub fn new_qualified(db: &dyn crate::Db, module: ModuleId, name: Symbol) -> Self {
        Self::new(db, Some(module), name)
    }
}
