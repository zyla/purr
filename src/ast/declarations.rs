use super::{Commented, Located, Pat, PossiblyGuardedExpr, Type};
use crate::ast::QualifiedName;
use crate::symbol::Symbol;

pub type Module = Located<Commented<ModuleInner>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ModuleInner {
    pub name: ModuleName,
    pub exports: Option<Vec<DeclarationRef>>,
    pub imports: Vec<Import>,
    pub declarations: Vec<Declaration>,
}

pub type Import = Located<ImportInner>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ImportInner {
    pub module: ModuleName,
    pub kind: ImportDeclarationKind,
    pub alias: Option<ModuleName>,
}

pub type DeclarationRef = Located<DeclarationRefKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum NameSource {
    UserNamed,
    CompilerNamed,
}

pub type ModuleName = QualifiedName;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationRefKind {
    TypeClass {
        name: Symbol,
    },
    TypeOp {
        name: Symbol,
    },
    Type {
        name: Symbol,
        constructors: Option<DeclarationRefConstructors>,
    },
    Value {
        name: Symbol,
    },
    ValueOp {
        name: Symbol,
    },
    TypeInstanceRef {
        name: Symbol,
        name_source: NameSource,
    },
    Module {
        name: ModuleName,
    },
    ReExport {
        imported_from: ModuleName,
        defined_in: ModuleName,
        declaration_ref: Box<DeclarationRef>,
    },
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationRefConstructors {
    All,
    Some(Vec<Symbol>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum ImportDeclarationKind {
    Implicit,
    Explicit(Vec<DeclarationRef>),
    Hiding(Vec<DeclarationRef>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct RoleDeclarationData {
    pub ident: Symbol,
    pub role: Vec<Role>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Role {
    Nominal,
    Representational,
    Phantom,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct TypeDeclarationData {
    pub ident: Symbol,
    pub r#type: Type,
}

impl TypeDeclarationData {
    pub fn new(ident: Symbol, r#type: Type) -> Self {
        Self { ident, r#type }
    }
}

pub type Declaration = Located<Commented<DeclarationKind>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DeclarationKind {
    Data {
        type_: DataDeclType,
        name: Symbol,
        params: Vec<TypeParameter>,
        kind: Option<Type>,
        constructors: Vec<DataConstructorDeclaration>,
    },

    // TODO: do we need this? seems internal
    DataBindingGroup(Vec<Declaration>),

    TypeSynonym {
        name: Symbol,
        params: Vec<TypeParameter>,
        body: Type,
    },

    KindSignature {
        for_type: KindSignatureFor,
        name: Symbol,
        kind: Kind,
    },

    Role(RoleDeclarationData),

    TypeSignature(TypeDeclarationData),

    ValueDeclaration(ValueDeclaration),

    // C x y = z, used in `let`
    Destructuring {
        pat: Pat,
        expr: PossiblyGuardedExpr,
    },

    ForeignValue {
        name: Symbol,
        type_: Type,
    },

    Class(TypeClassDeclaration),

    InstanceChain(Vec<InstanceDeclaration>),

    Operator {
        associativity: Associativity,
        precedence: u8,
        name: OperatorTarget,
        operator: Symbol,
    },
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Associativity {
    None,
    Left,
    Right,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum OperatorTarget {
    Type(QualifiedName),
    DataConstructor(QualifiedName),
    Value(QualifiedName),
}

pub type Kind = Type;

pub type TypeParameter = (Symbol, Option<Kind>);

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct TypeClassDeclaration {
    pub constraints: Vec<Type>,
    pub name: Symbol,
    pub params: Vec<TypeParameter>,
    pub fundeps: Vec<Fundep>,
    pub methods: Vec<TypeDeclarationData>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct InstanceDeclaration {
    pub constraints: Vec<Type>,
    pub instance_type: InstanceType,
    pub instance_name: Option<Symbol>,
    pub class: QualifiedName,
    pub args: Vec<Type>,
    pub body: Vec<Declaration>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct Fundep {
    pub from: Vec<Symbol>,
    pub to: Vec<Symbol>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum InstanceType {
    Plain,
    Derive,
    DeriveNewtype,
}

// Note: `data` and `newtype` signatures are actually declarations without constructors
#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum KindSignatureFor {
    TypeSynonym,
    Class,
}

pub type DataConstructorDeclaration = Located<Commented<DataConstructorDeclarationData>>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct DataConstructorDeclarationData {
    pub name: Symbol,

    // TODO: in original AST they have names. Why? I thought datacon fields are unnamed in PS
    pub fields: Vec<Type>,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DataDeclType {
    Data,
    ForeignData,
    Newtype,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct ValueDeclaration {
    pub ident: Symbol,

    // TODO: what is this?
    // pub name: NameKind
    pub params: Vec<Pat>,
    pub expr: PossiblyGuardedExpr,
}
