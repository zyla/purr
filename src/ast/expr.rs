use super::{Declaration, Located, Type};
use crate::ast::QualifiedName;
use crate::string::PSChar;
use crate::string::PSString;
use crate::symbol::Symbol;
use ordered_float::OrderedFloat;

pub type Expr = Located<ExprKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum ExprKind {
    Literal(Literal<Expr>),

    /// Infix operator sequence with unknown precedence
    Infix(Box<Expr>, Vec<(InfixOp, Expr)>),

    /// Record field accessor
    Accessor(Box<Expr>, Symbol),

    RecordUpdate(Box<Expr>, RecordUpdate),

    Var(QualifiedName),

    /// Standalone operator
    Operator(InfixOp),

    DataConstructor(QualifiedName),

    App(Box<Expr>, Vec<Expr>),

    Lam(Vec<Pat>, Box<Expr>),

    Case {
        exprs: Vec<Expr>,
        branches: Vec<CaseBranch>,
    },

    If {
        cond: Box<Expr>,
        then_: Box<Expr>,
        else_: Box<Expr>,
    },

    Typed(Box<Expr>, Box<Type>),

    Let {
        decls: Vec<Declaration>,
        body: Box<Expr>,
    },

    Wildcard,

    // Pseudo-expression, used only as an intermediate value during parsing.
    RecordUpdateSuffix(RecordUpdate),

    // Pseudo-expression, used only as an intermediate value during parsing.
    NamedPat(Symbol, Box<Expr>),

    Do(Vec<DoItem>),

    Ado(Vec<DoItem>, Box<Expr>),

    Negate(Box<Expr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum InfixOp {
    Symbol(QualifiedName),
    Backtick(Box<Expr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum DoItem {
    Let(Vec<Declaration>),
    Expr(Expr),
    Bind(Pat, Expr),
}

type RecordUpdate = Vec<(Symbol, Expr)>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum RecordLiteralOrUpdate {
    Literal(Vec<(Symbol, Expr)>),
    Update(Vec<(Symbol, Expr)>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct CaseBranch {
    pub pats: Vec<Pat>,
    pub expr: PossiblyGuardedExpr,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum PossiblyGuardedExpr {
    Unconditional(Expr),
    Guarded(Vec<GuardedExpr>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub struct GuardedExpr {
    pub guards: Vec<Guard>,
    pub expr: Expr,
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Guard {
    Expr(Expr),
    Bind(Pat, Expr),
}

pub type Pat = Located<PatKind>;

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum PatKind {
    Literal(Literal<Pat>),

    /// Infix operator sequence with unknown precedence
    Infix(Box<Pat>, Vec<(QualifiedName, Pat)>),

    Var(Symbol),

    DataConstructorApp(QualifiedName, Vec<Pat>),

    Wildcard,

    Named(Symbol, Box<Pat>),

    Typed(Box<Pat>, Box<Type>),
}

#[derive(Eq, PartialEq, Debug, Hash, Clone)]
pub enum Literal<T> {
    Integer(i64),
    Float(OrderedFloat<f64>),
    String(PSString),
    Char(PSChar),
    Boolean(bool),
    Array(Vec<T>),
    Object(Vec<(Symbol, T)>),
}

#[test]
fn test_size() {
    // Note: it was 56 before adding multi-case.
    assert_eq!(std::mem::size_of::<Expr>(), 72);
}
