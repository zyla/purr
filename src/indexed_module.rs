use crate::ast::AbsoluteName;
use crate::ast::CaseBranch;
use crate::ast::DataConstructorDeclaration;
use crate::ast::DataDeclType;
use crate::ast::Declaration;
use crate::ast::DeclarationKind;
use crate::ast::Fundep;
use crate::ast::Located;
use salsa::DebugWithDb;
use std::iter::Peekable;
use std::path::PathBuf;

use crate::ast::Type;
use crate::ast::TypeDeclarationData;
use crate::ast::TypeParameter;
use crate::ParsedModule;
use crate::{Db, ModuleId};
use crate::{Diagnostic, Diagnostics};
use fxhash::FxHashMap;
use std::collections::hash_map::Entry;

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub enum TypeDecl {
    Data(DataDecl),
    Type(TypeSynonymDecl),
    TypeClass(TypeClassDecl),
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct DataDecl {
    pub type_: DataDeclType,
    pub name: AbsoluteName,
    pub params: Vec<TypeParameter>,
    pub kind: Option<Type>,
    pub constructors: Vec<DataConstructorDeclaration>,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct TypeSynonymDecl {
    pub name: AbsoluteName,
    pub params: Vec<TypeParameter>,
    pub body: Type,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct TypeClassDecl {
    pub name: AbsoluteName,
    pub constraints: Vec<Type>,
    pub params: Vec<TypeParameter>,
    pub fundeps: Vec<Fundep>,
    pub methods: Vec<TypeDeclarationData>,
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct ValueDecl {
    pub name: AbsoluteName,
    pub type_: Option<Type>,

    /// Empty equations means foreign import.
    /// TODO: maybe make it more explicit?
    pub equations: Vec<CaseBranch>,
}

struct ModuleIndexer<'a> {
    db: &'a dyn Db,
    // Note: Using `FxHashMap` mostly because we want deterministic order for snapshots.
    types: FxHashMap<AbsoluteName, TypeDecl>,
    values: FxHashMap<AbsoluteName, ValueDecl>,
    classes: FxHashMap<AbsoluteName, TypeClassDecl>,
    module_id: ModuleId,
    filename: PathBuf,
    // TODO: Add map with Decl to SourceSpan
}

impl<'a> ModuleIndexer<'a> {
    fn index(&mut self, module: ParsedModule) {
        let db = self.db;
        let mut iter = module.ast.declarations.iter().peekable();
        while let Some(src_decl) = iter.peek().copied() {
            use DeclarationKind::*;

            let ref_loc = src_decl.0.start;

            match &***src_decl {
                Data {
                    type_,
                    name,
                    params,
                    kind,
                    constructors,
                } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);
                    let mut params = params.clone();
                    to_relative_params(abs_name, ref_loc, &mut params);

                    let mut constructors = constructors.clone();
                    constructors
                        .iter_mut()
                        .for_each(|c| c.to_relative_span(abs_name, ref_loc));

                    match self.types.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate type declaration {}", name.text(db)),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            e.insert(TypeDecl::Data(DataDecl {
                                type_: *type_,
                                name: abs_name,
                                params,
                                kind: kind.clone(),
                                constructors,
                            }));
                        }
                    }
                    iter.next();
                }

                TypeSynonym { name, params, body } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);

                    match self.types.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate type declaration {}", name.text(db)),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut params = params.clone();
                            to_relative_params(abs_name, ref_loc, &mut params);

                            let mut body = body.clone();
                            body.to_relative_span(abs_name, ref_loc);

                            e.insert(TypeDecl::Type(TypeSynonymDecl {
                                name: abs_name,
                                params,
                                body,
                            }));
                        }
                    }
                    iter.next();
                }

                KindSignature { .. } => todo!(),

                Role(_) => todo!(),

                TypeSignature(sig) => {
                    iter.next();
                    match iter.peek() {
                        Some(x) if matches!(&****x, ValueDeclaration(_)) => {
                            self.parse_value_decl(
                                &mut iter,
                                Some(Located::new(src_decl.span(), sig.clone())),
                                ref_loc,
                            );
                        }
                        _ => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!(
                                        "Type signature of {} should be followed by its definition",
                                        sig.ident.text(db)
                                    ),
                                    module.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                    }
                }
                ValueDeclaration(_) => {
                    self.parse_value_decl(&mut iter, None, ref_loc);
                }
                Destructuring { .. } => {
                    Diagnostics::push(
                        db,
                        Diagnostic::new(
                            src_decl.span().start,
                            src_decl.span().end,
                            "Invalid top-level destructuring".to_string(),
                            module.filename.to_string_lossy().into(),
                        ),
                    );
                    iter.next();
                }
                ForeignValue { name, type_ } => {
                    let abs_name = AbsoluteName::new(db, self.module_id, *name);

                    match self.values.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!("Duplicate value declaration {}", name.text(db)),
                                    self.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut type_ = type_.clone();
                            type_.to_relative_span(abs_name, ref_loc);

                            e.insert(ValueDecl {
                                name: abs_name,
                                type_: Some(type_),
                                equations: vec![],
                            });
                        }
                    }
                    iter.next();
                }
                Class(type_class_decl) => {
                    let abs_name = AbsoluteName::new(db, self.module_id, type_class_decl.name);

                    match self.classes.entry(abs_name) {
                        Entry::Occupied(_) => {
                            Diagnostics::push(
                                db,
                                Diagnostic::new(
                                    src_decl.span().start,
                                    src_decl.span().end,
                                    format!(
                                        "Duplicate typeclass declaration {}",
                                        type_class_decl.name.text(db)
                                    ),
                                    self.filename.to_string_lossy().into(),
                                ),
                            );
                        }
                        Entry::Vacant(e) => {
                            let mut constraints = type_class_decl.constraints.clone();
                            constraints
                                .iter_mut()
                                .for_each(|c| c.to_relative_span(abs_name, ref_loc));

                            let mut params = type_class_decl.params.clone();
                            to_relative_params(abs_name, ref_loc, &mut params);

                            let mut methods = type_class_decl.methods.clone();
                            methods
                                .iter_mut()
                                .for_each(|decl| decl.r#type.to_relative_span(abs_name, ref_loc));

                            e.insert(TypeClassDecl {
                                name: abs_name,
                                constraints,
                                params,
                                fundeps: type_class_decl.fundeps.clone(),
                                methods,
                            });
                        }
                    }

                    iter.next();
                }
                InstanceChain(_) => todo!(),
                Operator { .. } => todo!(),
            }
        }
    }

    fn parse_value_decl<'m>(
        &mut self,
        iter: &mut Peekable<impl Iterator<Item = &'m Declaration>>,
        sig: Option<Located<TypeDeclarationData>>,
        reference_loc: usize,
    ) {
        let db = self.db;
        let Some(first_decl) = iter.next() else {
            panic!("parse_value_decl should be called when next is ValueDeclaration")
        };
        let DeclarationKind::ValueDeclaration(first) = &***first_decl else {
            panic!("parse_value_decl should be called when next is ValueDeclaration")
        };

        let abs_name = AbsoluteName::new(db, self.module_id, first.ident);

        let ty = match sig {
            None => None,
            Some(sig) => {
                if sig.ident == first.ident {
                    let mut sig_ty = sig.into_inner().r#type;
                    sig_ty.to_relative_span(abs_name, reference_loc);
                    Some(sig_ty)
                } else {
                    Diagnostics::push(
                        db,
                        Diagnostic::new(
                            sig.span().start,
                            sig.span().end,
                            format!(
                                "Type signature of {} should be followed by its definition",
                                sig.ident.text(db)
                            ),
                            self.filename.to_string_lossy().into(),
                        ),
                    );
                    None
                }
            }
        };
        let mut equations = vec![first.clone().into()];

        //TODO: Make spans relative in equations SourceSpan
        while let Some(src_decl) = iter.peek() {
            let DeclarationKind::ValueDeclaration(decl) = &****src_decl else {
                break;
            };
            if decl.ident != first.ident {
                // Start of another chain of equations
                break;
            }
            equations.push(decl.clone().into());
            iter.next();
        }

        match self.values.entry(abs_name) {
            Entry::Occupied(_) => {
                Diagnostics::push(
                    db,
                    Diagnostic::new(
                        first_decl.span().start,
                        first_decl.span().end,
                        format!("Duplicate value declaration {}", first.ident.text(db)),
                        self.filename.to_string_lossy().into(),
                    ),
                );
            }
            Entry::Vacant(e) => {
                e.insert(ValueDecl {
                    name: abs_name,
                    type_: ty,
                    equations,
                });
            }
        }
    }
}

fn to_relative_params(
    abs_name: AbsoluteName,
    reference_loc: usize,
    params: &mut [(crate::symbol::Symbol, Option<Type>)],
) {
    params.iter_mut().for_each(|(_, type_opt)| {
        if let Some(typ) = type_opt {
            typ.to_relative_span(abs_name, reference_loc)
        }
    })
}

#[derive(PartialEq, Eq, Clone, Debug, DebugWithDb)]
pub struct IndexedModule {
    pub module_id: ModuleId,
    pub types: FxHashMap<AbsoluteName, TypeDecl>,
    pub values: FxHashMap<AbsoluteName, ValueDecl>,
    pub classes: FxHashMap<AbsoluteName, TypeClassDecl>,
}

#[salsa::tracked]
pub fn indexed_module(db: &dyn Db, module_id: ModuleId) -> IndexedModule {
    let module = crate::parsed_module(db, module_id);
    let mut indexer = ModuleIndexer {
        db,
        types: Default::default(),
        values: Default::default(),
        classes: Default::default(),
        module_id,
        filename: module.filename.clone(),
    };
    indexer.index(module);
    IndexedModule {
        module_id,
        types: indexer.types,
        values: indexer.values,
        classes: indexer.classes,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::tests::*;
    use indoc::indoc;
    use insta::{self, assert_snapshot};

    fn index_module(input: &str) -> String {
        let db = &mut crate::Database::test_single_file_db(input);
        let module_id = parse_module_id(input, db);
        format!(
            "{:#?}",
            (
                indexed_module(db, module_id).into_debug_all(db),
                indexed_module::accumulated::<Diagnostics>(db, module_id)
            )
        )
        .drop_salsa_id()
    }

    #[test]
    fn index_data_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        data Foo = Foo | Bar
        "
        )));
    }

    #[test]
    fn duplicate_data_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        data Foo
        data Foo
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_1() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_2() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        bar = 1
        "
        )));
    }

    #[test]
    fn type_sig_without_equations_3() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        data Foo
        "
        )));
    }

    #[test]
    fn value_decl() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo = 1
        "
        )));
    }

    #[test]
    fn value_decl_with_signature() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo :: Int
        foo = 1
        "
        )));
    }

    #[test]
    fn value_decl_many_equations() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo 1 = 1
        foo 2 = 2
        "
        )));
    }

    #[test]
    fn type_alias() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        type Foo = Int
        "
        )));
    }

    #[test]
    fn type_and_data_decl_conflict() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        type Foo = Int
        data Foo
        "
        )));
    }

    #[test]
    fn foreign_import() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foreign import foo :: Int
        "
        )));
    }

    #[test]
    fn duplicate_value() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        foo = 1
        bar = 2
        foo = 1
        "
        )));
    }

    #[test]
    fn typeclass() {
        assert_snapshot!(index_module(indoc!(
            "
        module Test where
        
        class Show a where
          show :: a -> String
        "
        )));
    }
}
