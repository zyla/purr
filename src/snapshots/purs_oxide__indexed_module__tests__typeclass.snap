---
source: src/indexed_module.rs
expression: "index_module(indoc!(\"\n        module Test where\n        \n        class Show a where\n          show :: a -> String\n        \"))"
---
(
    IndexedModule {
        module_id: ModuleId {
            name: "Test",
        },
        types: {},
        values: {},
        classes: {
            AbsoluteName {
                module: ModuleId {
                    name: "Test",
                },
                name: Symbol {
                    text: "Show",
                },
            }: TypeClassDecl {
                name: AbsoluteName {
                    module: ModuleId {
                        name: "Test",
                    },
                    name: Symbol {
                        text: "Show",
                    },
                },
                constraints: [],
                params: [
                    (
                        Symbol {
                            text: "a",
                        },
                        None,
                    ),
                ],
                fundeps: [],
                methods: [
                    TypeDeclarationData {
                        ident: Symbol {
                            text: "show",
                        },
                        r#type: Located(
                            SourceSpan {
                                decl: Decl(
                                    DeclId {
                                        namespace: Class,
                                        module: ModuleId {
                                            name: "Test",
                                        },
                                        name: Symbol {
                                            text: "Show",
                                        },
                                    },
                                ),
                                start: 29,
                                end: 40,
                            },
                            FunctionType(
                                Located(
                                    SourceSpan {
                                        decl: Module(
                                            ModuleId {
                                                name: "Test",
                                            },
                                        ),
                                        start: 48,
                                        end: 49,
                                    },
                                    Var(
                                        Symbol {
                                            text: "a",
                                        },
                                    ),
                                ),
                                Located(
                                    SourceSpan {
                                        decl: Module(
                                            ModuleId {
                                                name: "Test",
                                            },
                                        ),
                                        start: 53,
                                        end: 59,
                                    },
                                    TypeConstructor(
                                        QualifiedName {
                                            module: None,
                                            name: Symbol {
                                                text: "String",
                                            },
                                        },
                                    ),
                                ),
                            ),
                        ),
                    },
                ],
            },
        },
        decls_ref_loc: {
            DeclId(
                Id {
                    value: 1,
                },
            ): 19,
        },
        filename: "test.purs",
    },
    [],
)
